### PR TITLE
Optimize DateTimeFormatter format & parse

### PR DESCRIPTION
**Pull request by wenshao**
This PR optimizes the performance of DateTimeFormatter.format. The optimization work includes the following:

## 1. Reduce method size to be inlined by C2
Refactor the NumberPrinterParser.format method. Currently, the codeSize of NumberPrinterParser.format is > 325 and cannot be inlined by the C2 optimizer.

```
java.time.format.DateTimeFormatterBuilder$NumberPrinterParser::format (395 bytes)
```

The exception handling code can be extracted into a separate method to reduce the codeSize of the method, as follows:

```java
        final void maxWidthError(long value) {
            throw new DateTimeException("Field " + field +
                    " cannot be printed as the value " + value +
                    " exceeds the maximum print width of " + maxWidth);
        }

        final void negateFieldError(long value) {
            throw new DateTimeException("Field " + field +
                    " cannot be printed as the value " + value +
                    " cannot be negative according to the SignStyle");
        }

```

## 2. Provide targeted optimization implementation for common scenarios

Provides optimized NumberPrinterParser subclasses for common field patterns, covering common patterns, including:
```
M
MM
d
dd
Y
YYYY
W
w
H
HH
h
hh
m
mm
s
ss
SSS
```

```java
    static class NumberPrinterParser implements DateTimePrinterParser {
        static NumberPrinterParser of(TemporalField field, int minWidth, int maxWidth, SignStyle signStyle, int subsequentWidth) {
            // Subclasses optimized for specific scenarios
            var range = field.range();
            long minimum    = range.getMinimum(),
                 maximum    = range.getMaximum();
            int maximumSize = DecimalDigits.stringSize(maximum),
                minimumSize = DecimalDigits.stringSize(Math.abs(minimum));
            if (signStyle == SignStyle.NOT_NEGATIVE) {
                if (minWidth == 1 && maxWidth == 1 && minimum >= 0 && maximumSize == 1) {
                    return new Width1NotNegative(field, subsequentWidth);
                }
                if (maxWidth == 2 && minimum >= 0) {
                    if (minimum >= 0 && maximumSize <= 2) {
                        if (minWidth == 1) {
                            return new Width2NotNegative(field, subsequentWidth);
                        }
                        if (minWidth == 2) {
                            return new FixWidth2NotNegative(field, subsequentWidth);
                        }
                    }
                }
                if (minWidth == 3 && maxWidth == 3 && minimum >= 0 && maximumSize <= 3) {
                    return new FixWidth3NotNegative(field, subsequentWidth);
                }
                if (minWidth == 4 && maxWidth == 4) {
                    // nedd range check
                    return new FixWidth4NotNegative(field, subsequentWidth);
                }
            }
            if (minWidth == 4) {
                if (signStyle == SignStyle.EXCEEDS_PAD) {
                    if (minimumSize <= maxWidth && maximumSize <= maxWidth) {
                        return new Width4ExceedsPad(field, maxWidth);
                    }
                }
            }
            return new NumberPrinterParser(field, minWidth, maxWidth, signStyle, subsequentWidth);
        }

        static final class FixWidth2NotNegative extends NumberPrinterParser {
            private FixWidth2NotNegative(TemporalField field, int subsequentWidth) {
                super(field, 2, 2, SignStyle.NOT_NEGATIVE, subsequentWidth);
            }
            @Override
            public void format(StringBuilder buf, DecimalStyle decimalStyle, long value) {
                int val = (int) value;
                char zeroDigit = decimalStyle.getZeroDigit();
                buf.append((char) (zeroDigit + val / 10))
                   .append((char) (zeroDigit + val % 10));
            }
        }
    }

```

## 3. Convert printerParsers of CompositePrinterParser to concrete fields so that they can be optimized by TypeProfile in C2
Use subclasses with printerParsers.length of 1 to 15 so that C2 can do TypeProfile optimization

``` java
	static class CompositePrinterParser implements DateTimePrinterParser {
        static CompositePrinterParser of(DateTimePrinterParser[] printerParsers, boolean optional) {
            // Use subclasses with printerParsers.length of 1 to 15 so that C2 can do TypeProfile optimization
            return switch (printerParsers.length) {
                case 1 -> new N1(printerParsers, optional);
                case 2 -> new N2(printerParsers, optional);
                case 3 -> new N3(printerParsers, optional);
                case 4 -> new N4(printerParsers, optional);
                case 5 -> new N5(printerParsers, optional);
                case 6 -> new N6(printerParsers, optional);
                case 7 -> new N7(printerParsers, optional);
                case 8 -> new N8(printerParsers, optional);
                case 9 -> new N9(printerParsers, optional);
                case 10 -> new N10(printerParsers, optional);
                case 11 -> new N11(printerParsers, optional);
                case 12 -> new N12(printerParsers, optional);
                case 13 -> new N13(printerParsers, optional);
                case 14 -> new N14(printerParsers, optional);
                case 15 -> new N15(printerParsers, optional);
                default -> new CompositePrinterParser(printerParsers, optional);
            };
        }

        static class N1 extends CompositePrinterParser {
            final DateTimePrinterParser p0;
        }

        static class N2 extends N1 {
            final DateTimePrinterParser p1;
            protected N2(DateTimePrinterParser[] printerParsers, boolean optional) {
                super(printerParsers, optional);
                p1 = printerParsers[1];
            }

            public boolean format0(DateTimePrintContext context, StringBuilder buf) {
                return p0.format(context, buf)
                    && p1.format(context, buf);
            }
        }

        static class N2 extends N1 {
            final DateTimePrinterParser p1;
            protected N2(DateTimePrinterParser[] printerParsers, boolean optional) {
                super(printerParsers, optional);
                p1 = printerParsers[1];
            }
            public boolean format0(DateTimePrintContext context, StringBuilder buf) {
                return p0.format(context, buf)
                    && p1.format(context, buf);
            }
        }

        // ...

        static class N15 extends N14 {
            final DateTimePrinterParser p14;
            protected N15(DateTimePrinterParser[] printerParsers, boolean optional) {
                super(printerParsers, optional);
                p14 = printerParsers[14];
            }
            public boolean format0(DateTimePrintContext context, StringBuilder buf) {
                return p0.format(context, buf)
                    && p1.format(context, buf)
                    && p2.format(context, buf)
                    && p3.format(context, buf)
                    && p4.format(context, buf)
                    && p5.format(context, buf)
                    && p6.format(context, buf)
                    && p7.format(context, buf)
                    && p8.format(context, buf)
                    && p9.format(context, buf)
                    && p10.format(context, buf)
                    && p11.format(context, buf)
                    && p12.format(context, buf)
                    && p13.format(context, buf)
                    && p14.format(context, buf);
            }
        }
	}

```

We can see that TypeProfile works

```
 @ 6   java.time.format.DateTimeFormatterBuilder$NumberPrinterParser::format (35 bytes)   inline (hot)
  \-> TypeProfile (15360/15360 counts) = java/time/format/DateTimeFormatterBuilder$NumberPrinterParser$Width4ExceedsPad

 @118   java.time.format.DateTimeFormatterBuilder$NumberPrinterParser::format (35 bytes)   inline (hot)
  \-> TypeProfile (15360/15360 counts) = java/time/format/DateTimeFormatterBuilder$NumberPrinterParser$FixWidth2NotNegative
```

## 4. Special optimization for commonly used patterns
Provides a monolithic NumberPrinterParser implementation for commonly used patterns, similar to InstantPrinterParser, instead of using CompositePrinterParser.
```
HH:mm:ss
HH:mm:ss.SSS
yyyy-MM-dd'T'HH:mm:ss
yyyy-MM-dd'T'HH:mm:ss.SSS这些
```

```java
public final class DateTimeFormatterBuilder {
	static class CompositePrinterParser implements DateTimePrinterParser {
	
	    /*
	     * yyyy-MM-dd
	     */
		static final class YMD extends N5 {
			final NumberPrinterParser.Width4ExceedsPad printerParserYear;
            final NumberPrinterParser.FixWidth2NotNegative printerParserMonth;
            final NumberPrinterParser.FixWidth2NotNegative printerParserDayOfMonth;

            ...

            @Override
            public boolean format(DateTimePrintContext context, StringBuilder buf) {
                var temporal = context.getTemporal();
                int year       = temporal.getYear();
                int month      = temporal.getMonthValue();
                int dayOfMonth = temporal.getDayOfMonth();

                if (printerParserYear.field == ChronoField.YEAR_OF_ERA) {
                    year = (year >= 1 ? year : 1 - year);
                }

                printerParserYear.format(buf, context.getDecimalStyle(), year);
                var zeroDigit = context.getDecimalStyle().getZeroDigit();
                buf.append(literal)
                   .append((char) (zeroDigit + month / 10))
                   .append((char) (zeroDigit + month % 10))
                   .append(literal)
                   .append((char) (zeroDigit + dayOfMonth / 10))
                   .append((char) (zeroDigit + dayOfMonth % 10));
                return true;
            }

            public String format(TemporalAccessor temporal, DateTimeFormatter formatter) {
                char zeroDigit = formatter.getDecimalStyle().getZeroDigit();
                if (zeroDigit != '0'
                        && (literal >>> 8 == 0) // StringLatin1.canEncode
                ) {
                    StringBuilder buf = new StringBuilder(15);
                    format(new DateTimePrintContext(temporal, formatter), buf);
                    return buf.toString();
                }

                int year       = temporal.getYear();
                int month      = temporal.getMonthValue();
                int dayOfMonth = temporal.getDayOfMonth();

                int size   = printerParserYear.size(year) + 6;
                byte[] buf = new byte[size];
                int index  = printerParserYear.formatTo(buf, 0, year);
                buf[index] = (byte) literal;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, month);
                buf[index] = (byte) literal;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, dayOfMonth);
                try {
                    return JLA.newStringNoRepl(buf, StandardCharsets.ISO_8859_1);
                } catch (CharacterCodingException cce) {
                    throw new AssertionError(cce);
                }
            }

            public <T> T parse(CharSequence text, DateTimeFormatter formatter, TemporalQuery<T> query) {
                char zeroDigit = formatter.getDecimalStyle().getZeroDigit();
                if (zeroDigit != '0') {
                    return super.parse(text, formatter, query);
                }

                long valuePos = printerParserYear.parse(text, 0);
                int pos = position(text, valuePos);
                int year = (int) (valuePos >> 32);

                pos = litteral(text, pos, literal);

                valuePos = printerParserMonth.parse(text, pos);
                pos = position(text, valuePos);
                int month = (int) (valuePos >> 32);

                pos = litteral(text, pos, literal);

                valuePos = printerParserDayOfMonth.parse(text, pos);
                position(text, valuePos);
                int dayOfMonth = (int) (valuePos >> 32);

                return query.queryFrom(
                        LocalDate.of(year, month, dayOfMonth));
            }
            ...
		}
	}
}
```
This optimization reduces many intermediate steps, such as the very time-consuming DateTimeParseContext, thereby further improving performance.

## 5. Provide a version based on latin1 encoding byte[]
```java
	static class CompositePrinterParser implements DateTimePrinterParser {	
		static final class Width4ExceedsPad extends NumberPrinterParser {
            public int size(int value) {
                if (value > -1000 && value < 10000) {
                    return 4;
                }

                int size = 0;
                if (value >= 0) {
                    if (value > 9999) {
                        size = 1;
                    }
                } else {
                    value = -value;
                    size = 1;
                }
                return size + DecimalDigits.stringSize(value);
            }

            public int formatTo(byte[] buf, int index, int value) {
                if (value >= 0) {
                    if (value > 9999) {
                        buf[index++] = '+';
                    }
                } else {
                    value = -value;
                    buf[index++] = '-';
                }
                if (value < 10000) {
                    int div = value / 100;
                    int rem = value % 100;
                    index = DecimalDigits.putPairLatin1(buf, index, div);
                    index = DecimalDigits.putPairLatin1(buf, index, rem);
                } else {
                    index += DecimalDigits.stringSize(value);
                    JLA.getCharsLatin1(value, index, buf);
                }
                return index;
            }
		}

		static final class YMDHMS extends N11 {
			final NumberPrinterParser.Width4ExceedsPad     printerParserYear;

            public String format(TemporalAccessor temporal, DateTimeFormatter formatter) {
                var zeroDigit = formatter.getDecimalStyle().getZeroDigit();
                if (zeroDigit != '0'
                        && (literal1 >>> 8 == 0) // StringLatin1.canEncode
                        && (literal5 >>> 8 == 0) // StringLatin1.canEncode
                        && (literal7 >>> 8 == 0) // StringLatin1.canEncode
                ) {
                    StringBuilder buf = new StringBuilder(32);
                    format(new DateTimePrintContext(temporal, formatter), buf);
                    return buf.toString();
                }

                int year       = temporal.getYear();
                int month      = temporal.getMonthValue();
                int dayOfMonth = temporal.getDayOfMonth();
                int hour       = temporal.getHour();
                int minute     = temporal.getMinute();
                int second     = temporal.getSecond();

                int size = printerParserYear.size(year) + 15;
                byte[] buf = new byte[size];
                int index  = printerParserYear.formatTo(buf, 0, year);
                buf[index] = (byte) literal1;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, month);
                buf[index] = (byte) literal1;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, dayOfMonth);
                buf[index] = (byte) literal5;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, hour);
                buf[index] = (byte) literal7;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, minute);
                buf[index] = (byte) literal7;
                index      = DecimalDigits.putPairLatin1(buf, index + 1, second);
                try {
                    return JLA.newStringNoRepl(buf, StandardCharsets.ISO_8859_1);
                } catch (CharacterCodingException cce) {
                    throw new AssertionError(cce);
                }
            }
        }
    }
```

## 6. Summary

This PR uses a lot of optimizations and has a significant performance improvement, but there are too many changes and it will not be submitted as a formal PR. We can discuss it here and finally split it into multiple small PRs.

3. The same optimization applies to parse

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20391/head:pull/20391` \
`$ git checkout pull/20391`

Update a local copy of the PR: \
`$ git checkout pull/20391` \
`$ git pull https://git.openjdk.org/jdk.git pull/20391/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20391`

View PR using the GUI difftool: \
`$ git pr show -t 20391`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20391.diff">https://git.openjdk.org/jdk/pull/20391.diff</a>

</details>

### CODE REVIEW

**Comment by liach** [comment]
@wenshao Since you said that you have tried a bytecode approach, can you share your code that you later found inefficient and link it, so I can check why it was slow?

File: src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java [hunk 1]
```diff
@@ -2502,12 +2503,34 @@ interface DateTimePrinterParser {
...
      * Composite printer and parser.
      */
-    static final class CompositePrinterParser implements DateTimePrinterParser {
+    static class CompositePrinterParser implements DateTimePrinterParser {
         private final DateTimePrinterParser[] printerParsers;
-        private final boolean optional;
+        final boolean optional;
+
+        static CompositePrinterParser of(DateTimePrinterParser[] printerParsers, boolean optional) {
+            // Use subclasses with printerParsers.length of 1 to 15 so that C2 can do TypeProfile optimization
```
[the remark this answers is not in the document]
Comment (wenshao) [reply src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java:2547]: Annotated `@Stable` TypeProfile still does not work

**Comment by wenshao** [comment]
@liach 
It will be faster based on bytecode. I modified the description of the PR and the bytecode implementation has been submitted.

**Comment by wenshao** [comment]
```java
static final MethodTypeDesc MTD_void_int = MethodTypeDesc.of(CD_void, CD_int);
static final MethodTypeDesc MTD_boolean  = MethodTypeDesc.of(CD_boolean);
static final MethodTypeDesc MTD_int      = MethodTypeDesc.of(CD_int);
static final MethodTypeDesc MTD_long     = MethodTypeDesc.of(CD_long);
 ```
@liach 
These should be added to ConstantDescs, if you think this is ok, I'll create a PR

**Comment by wenshao** [comment]
Here are the performance numbers running on a MacBook M1, It can be seen that the performance of both bytecode-based and non-bytecode-based improvements has been significantly improved.

``` bash
make test TEST="micro:java.time.format.DateTimeFormatterBench"

# non-bytecode
make test TEST="micro:java.time.format.DateTimeFormatterBench" MICRO="VM_OPTIONS=-Djava.time.format.DateTimeFormatter.compile=false"

```

```diff
-# master 5ff7c57f9ff5428ef3d2aedd7e860bb1e8ff29ea
-Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units
-DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  16.481 ? 0.211  ops/ms
-DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  10.818 ? 0.092  ops/ms
-DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15   8.830 ? 0.178  ops/ms
-DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15   6.760 ? 0.175  ops/ms
-DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  23.618 ? 0.155  ops/ms
-DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  16.038 ? 0.087  ops/ms
-DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  13.885 ? 0.043  ops/ms
-DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  10.144 ? 0.094  ops/ms

+# current 667d7f8f6f4345b949b4b7b79bfefe3e00115e07
+Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units
+DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  31.052 ? 0.124  ops/ms
+DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  26.573 ? 1.951  ops/ms
+DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15  21.599 ? 0.049  ops/ms
+DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  18.301 ? 0.187  ops/ms
+DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  84.841 ? 0.689  ops/ms
+DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  61.124 ? 0.580  ops/ms
+DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  44.280 ? 0.137  ops/ms
+DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  35.753 ? 0.224  ops/ms

+# current 667d7f8f6f4345b949b4b7b79bfefe3e00115e07
+# -Djava.time.format.DateTimeFormatter.compile=false
+Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units
+DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  31.291 ? 0.084  ops/ms
+DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  23.474 ? 0.097  ops/ms
+DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15  19.013 ? 0.061  ops/ms
+DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  15.431 ? 0.189  ops/ms
+DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  75.275 ? 0.065  ops/ms
+DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  54.709 ? 0.080  ops/ms
+DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  37.073 ? 0.068  ops/ms
+DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  31.344 ? 0.030  ops/ms
```

|   | pattern | baseline  | current | current-non-bytecode | delta |delta-non-bytecode |
| --- | --- | --- | --- | --- | --- | --- |
| formatInstants | HH:mm:ss | 16.481 | 31.052 | 31.291 | 88.41% | 89.86% |
| formatInstants | HH:mm:ss.SSS | 10.818 | 26.573 | 23.474 | 145.64% | 116.99% |
| formatInstants | yyyy-MM-dd'T'HH:mm:ss | 8.830 | 21.599 | 19.013 | 144.61% | 115.32% |
| formatInstants | yyyy-MM-dd'T'HH:mm:ss.SSS | 6.760 | 18.301 | 15.431 | 170.72% | 128.27% |
| formatZonedDateTime | HH:mm:ss | 23.618 | 84.841 | 75.275 | 259.22% | 218.72% |
| formatZonedDateTime | HH:mm:ss.SSS | 16.038 | 61.124 | 54.709 | 281.12% | 241.12% |
| formatZonedDateTime | yyyy-MM-dd'T'HH:mm:ss | 13.885 | 44.280 | 37.073 | 218.91% | 167.00% |
| formatZonedDateTime | yyyy-MM-dd'T'HH:mm:ss.SSS | 10.144 | 35.753 | 31.344 | 252.45% | 208.99% |

**Comment by wenshao** [comment]
Special optimizations have been made for common Patterns, including:
```
yyyy-MM-dd
HH:mm:ss
yyyy-MM-dd`T`HH:mm:ss
yyyy-MM-dd`T`HH:mm:ss.SSS
```

Performance has been improved again, and the performance numbers under MacBook M1 Pro are as follows:

```diff
-# baseline 5ff7c57f9ff5428ef3d2aedd7e860bb1e8ff29ea
-Benchmark                        Mode  Cnt     Score      Error   Units
-DateTimeFormatterParse.parse10  thrpt   15  6374.420 ? 1817.921  ops/ms
-DateTimeFormatterParse.parse19  thrpt   15  4315.269 ?   53.650  ops/ms
-DateTimeFormatterParse.parse23  thrpt   15  3898.511 ?   84.479  ops/ms

+# current 69aa432a056dbe9085dda9f757e87d105b5db4c9
+Benchmark                        Mode  Cnt      Score      Error   Units
+DateTimeFormatterParse.parse10  thrpt   15  59394.564 ? 2159.241  ops/ms
+DateTimeFormatterParse.parse19  thrpt   15  31054.803 ?  439.493  ops/ms
+DateTimeFormatterParse.parse23  thrpt   15  12397.622 ?  452.563  ops/ms
```
|   | baseline  | current | delta |
| --- | --- | --- | --- |
| DateTimeFormatterParse.parse10 | 6374.420 | 59394.564 | 831.76% |
| DateTimeFormatterParse.parse19 | 4315.269 | 31054.803 | 619.65% |
| DateTimeFormatterParse.parse23 | 3898.511 | 12397.622 | 218.01% |


```diff
-# baseline 5ff7c57f9ff5428ef3d2aedd7e860bb1e8ff29ea
-Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units
-DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  16.481 ? 0.211  ops/ms
-DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  10.818 ? 0.092  ops/ms
-DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15   8.830 ? 0.178  ops/ms
-DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15   6.760 ? 0.175  ops/ms
-DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  23.618 ? 0.155  ops/ms
-DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  16.038 ? 0.087  ops/ms
-DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  13.885 ? 0.043  ops/ms
-DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  10.144 ? 0.094  ops/ms

+# current 69aa432a056dbe9085dda9f757e87d105b5db4c9
+Benchmark                                                   (pattern)   Mode  Cnt   Score   Error   Units
+DateTimeFormatterBench.formatInstants                        HH:mm:ss  thrpt   15  37.119 ? 0.177  ops/ms
+DateTimeFormatterBench.formatInstants                    HH:mm:ss.SSS  thrpt   15  25.391 ? 0.042  ops/ms
+DateTimeFormatterBench.formatInstants           yyyy-MM-dd'T'HH:mm:ss  thrpt   15  25.186 ? 2.403  ops/ms
+DateTimeFormatterBench.formatInstants       yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  20.529 ? 4.622  ops/ms
+DateTimeFormatterBench.formatZonedDateTime                   HH:mm:ss  thrpt   15  89.618 ? 0.720  ops/ms
+DateTimeFormatterBench.formatZonedDateTime               HH:mm:ss.SSS  thrpt   15  61.136 ? 0.116  ops/ms
+DateTimeFormatterBench.formatZonedDateTime      yyyy-MM-dd'T'HH:mm:ss  thrpt   15  46.283 ? 0.175  ops/ms
+DateTimeFormatterBench.formatZonedDateTime  yyyy-MM-dd'T'HH:mm:ss.SSS  thrpt   15  38.452 ? 0.066  ops/ms
```

|   | pattern | baseline  | current | delta |
| --- | --- | --- | --- | --- |
| DateTimeFormatterBench.formatInstants | HH:mm:ss | 16.481 | 37.119 | 125.22% |
| DateTimeFormatterBench.formatInstants | HH:mm:ss.SSS | 10.818 | 25.391 | 134.71% |
| DateTimeFormatterBench.formatInstants | yyyy-MM-dd'T'HH:mm:ss | 8.830 | 25.186 | 185.23% |
| DateTimeFormatterBench.formatInstants | yyyy-MM-dd'T'HH:mm:ss.SSS | 6.760 | 20.529 | 203.68% |
| DateTimeFormatterBench.formatZonedDateTime | HH:mm:ss | 23.618 | 89.618 | 279.45% |
| DateTimeFormatterBench.formatZonedDateTime | HH:mm:ss.SSS | 16.038 | 61.136 | 281.19% |
| DateTimeFormatterBench.formatZonedDateTime | yyyy-MM-dd'T'HH:mm:ss | 13.885 | 46.283 | 233.33% |
| DateTimeFormatterBench.formatZonedDateTime | yyyy-MM-dd'T'HH:mm:ss.SSS | 10.144 | 38.452 | 279.06% |

**Comment by wenshao** [comment]
New PR #20559 has been submitted

**Pull request closed without merging**